メッセージ送信機能の非同期通信化

ars-ishi opened the pull request:
# WHAT
メッセージ送信機能の非同期通信化する

# WHY
非同期化しない場合、
メッセージを送信するたびに送信画面にリダイレクトしてビューが毎回再描画されてしまうため。

ars-ishi committed:
first commit

ars-ishi committed:
Edit messages/index.html.haml

messages/index.html.hamlを編集

ars-ishi committed:
Edit messages/_group-area.html.haml

groups/:group_id/messages に遷移するときのturbolinksを無効にした

ars-ishi committed:
Edit messages_controller.rb

respond_toを使用してHTMLとJSONの場合で処理を分けた。

ars-ishi committed:
Create messages/create.json.jbuilder

返すデータを作成した。

ars-ishi committed:
Edit messages/_chat-area__form.html.haml

メッセージのformにidをつけた。

ars-ishi committed:
Create message.js

・フォームが送信されたとき、入力された内容でHTMLを作成し、一番下に追加
・追加した分、メッセージ画面を下にスクロール
・失敗したときエラーメッセージを表示

ars-ishi commented:
ビューの動きです。
[![Image from Gyazo](https://i.gyazo.com/4e294b836fffb7e5f189d81b2a3d05ea.gif)](https://gyazo.com/4e294b836fffb7e5f189d81b2a3d05ea)

ars-ishi commented:
メッセージ送信機能を非同期通信化しました。ご確認お願いいたします。

kohei-isoda commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,53 @@
+$(function() {
+  function buildHTML(message){
+    if (message.image !== null) {
```
この書き方ですとhtmlの記述を2回書かなくてはなりません。
imageという変数（仮）を作って、画像が空の時にはnullが、画像がある時には画像が入るようにして、
その変数を使えばさらにシンプルに記述が出来そうです！ご検討くださいませ。

ars-ishi replied:
imageという変数を作って、画像が空の時にはnullが、画像がある時には画像が入るようにする形で記述を修正しました。

kohei-isoda replied:
修正ありがとうございます！
三項演算子というものを使えば、ワンライナー（1行）で書くことができますので調べてみてください！

kohei-isoda commented on app/views/messages/create.json.jbuilder:
```diff
@@ -0,0 +1,6 @@
+json.group_id @message.group.id
+json.user_id @message.user.id
+json.user_name @message.user.name
+json.created_at @message.created_at.strftime('%Y/%m/%d %H:%M')
```
json.created_atとなっていますが、created_atよりもフォーマットを加工しているのでdateなどにした方が分かりやすいかもしれません。

また、"rails time format"と検索していただくと下記のような記事が出て来ます。今よりもっとシンプルに書くことができますので参考にしてみてください！
https://qiita.com/Vit-Symty/items/399c77d1fd681b77d593

ars-ishi replied:
・json.created_atの名称をjson.dateに変更しました。
・time_formats.rbを作成して呼び出すように変更しました。

kohei-isoda commented on app/views/messages/create.json.jbuilder:
```diff
@@ -0,0 +1,6 @@
+json.content @message.content
+json.image @message.image_url
+json.group_id @message.group.id
```
group_idはjsonで使われていますか？使っていないのであれば不要かと！

ars-ishi replied:
group_idの記述を削除しました。

kohei-isoda commented on app/views/messages/create.json.jbuilder:
```diff
@@ -0,0 +1,6 @@
+json.content @message.content
+json.image @message.image_url
+json.group_id @message.group.id
+json.user_id @message.user.id
```
idはファイルの一番上に書いた方が可読性上がります！

ars-ishi replied:
json.user_idが一番上にくるように順序を変更しました。

kohei-isoda reviewed (commented):
修正のほど、宜しくお願い致します！

ars-ishi committed:
Create time_formats.rb

日時表示のデフォルトフォーマットを作成した。

ars-ishi committed:
Edit messages/create.json.jbuilder

・group_idは使用しなかったので削除
・idが上にくるように順序を変更
・日時データの名称と取得の仕方を変更

ars-ishi committed:
Edit javascripts/message.js

Htmlの記述が一度で済むようにimage変数を定義

ars-ishi commented:
ご指摘いただいた箇所を修正しました。ご確認お願いいたします。
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219730639
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219730763
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219730816
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219730836

kohei-isoda commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,50 @@
+                  <div class="message-list__name">${message.user_name}</div>
+                  <div class="message-list__date">${message.date}</div>
+                  <div class="message-list__text">${message.content}</div>`
+                  + image +
```
divタグなどは残しておいて良いと思います！

ars-ishi replied:
・var imageの条件分岐を三項演算子で書くようにしました。
・条件分岐にはimgだけ残してdivはhtmlの方に入れるようにしました。

kohei-isoda commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,50 @@
+                  <div class="message-list__text">${message.content}</div>`
+                  + image +
+                `</li>`
+    console.log(JSON.stringify(message)); //データ確認用
```
不要なconsole.logは削除しましょう！

ars-ishi replied:
console.logと、別の場所にコメントも入れていたので削除しました。

kohei-isoda commented on app/views/messages/_group-area.html.haml:
```diff
@@ -12,7 +12,7 @@
     - current_user.groups.each do |group|
       %li
-        = link_to group_messages_path(group) do
+        = link_to group_messages_path(group), data: { turbolinks: false } do
```
turbolinks: falseにしている理由はございますか？

ars-ishi replied:
こちらのリンクからgroups/id/messagesに遷移した直後にメッセージを投稿すると、
初回の投稿だけ投稿後のスクロールが発生しなかったため、
 turbolinksで遷移したことが要因かと思い、turbolinks: falseにしていました。
また、turbolinks: falseにした後はこちらのリンクから遷移してもスクロールするようになりました。

こちらの記述は削除し、
application.html.hamlとjavascripts/application.jsでturbolinksの設定を外すように変更しました。

kohei-isoda commented on app/views/messages/create.json.jbuilder:
```diff
@@ -0,0 +1,5 @@
+json.user_id @message.user.id
+json.user_name @message.user.name
+json.date Time.now.to_s
```
@message.created_at.to_s(:datetime)で良いかと！

ars-ishi replied:
Time.now.to_s　→　@message.created_at.to_s(:datetime)に修正いたしました。

kohei-isoda reviewed (commented):
再提出お願い致します！

ars-ishi committed:
Edit messages/create.json.jbuilder

json.dateのデータの記述の仕方を変更しました。

ars-ishi committed:
Edit javascripts/message.js

・message.imageの条件分岐を一行で書くように修正
・条件分岐では<img src="${message.image}">だけ残すように修正
・デバッグ用のconsole.logを削除
・コメントを削除

ars-ishi committed:
Edit _group-area.html.haml

・data: { turbolinks: false }を削除
・（turbolinks自体を使用しないように別ファイルで設定）

ars-ishi committed:
Edit gemfile

gem 'turbolinks', '~> 5'　をコメントアウト

ars-ishi committed:
Edit application.html.haml

stylesheet_link_tag と javascript_include_tagから
'data-turbolinks-track': 'reload’を削除

ars-ishi committed:
Edit javascripts/application.js

//= require turbolinks　を削除

ars-ishi commented:
ご指摘いただいた箇所を修正しました。ご確認お願いいたします。
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219753856
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219754311
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219754423
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219754778
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219755037

miyagit commented on app/views/messages/index.html.haml:
```diff
@@ -4,9 +4,5 @@
-        - else
-          %li
-            .message-list__text まだメッセージはありません。メッセージや画像を投稿してみましょう。
+        = render partial: "chat-area__timeline", collection: @messages, as: "message"
```
render @messagesのほうが良い気がしますね！
沢山部分テンプレートがあって使い分けたいならpartial: "chat-area__timeline"でも良いかもしれないですが。。

ars-ishi replied:
部分テンプレートの名称を変更してrender @messagesで呼ぶようにしました。

miyagit commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,45 @@
+        var html = buildHTML(data);
+        $('.timeline__message-list').append(html);
+        $('.form__field__text').val('');
+        $('.form__field__image').val('');
```
form属性のプロパティを空にするメソッド使いましょう。

ars-ishi replied:
text_fieldとfile_fieldそれぞれでvalを空にしていた部分を、
form自体をリセットするように修正しました。

ars-ishi committed:
Edit javascripts/message.js

text_fieldとfile_fieldそれぞれでvalを空にしていた部分を、
form自体をリセットするように修正しました。

ars-ishi committed:
Edit messages/index.html.haml

renderの記述を簡略化しました。

ars-ishi committed:
Rename from(_chat-area__timeline) to (message)

部分テンプレートの名称を変更しました。

ars-ishi commented:
ご指摘頂いた部分を修正しました。ご確認のほどお願いいたします。
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219778375
https://github.com/ars-ishi/chat-space/pull/11#discussion_r219778484

kohei-isoda commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,44 @@
+$(function() {
+  function buildHTML(message){
+    var image = message.image !== null ? `<img src="${message.image}">` : "";
```
nullでなければという条件ではなくmessage.imageがあればというシンプルな式の方が読みやすいです！

ars-ishi replied:
条件部分の書き方を修正しました。
message.image !== null
↓
message.image

kohei-isoda reviewed (commented):
一点だけ細かいですが、コメント致しました！

ars-ishi committed:
Edit javascripts/message.js

三項演算子の条件部分の書き方を修正しました。
message.image !== null
↓
message.image

ars-ishi commented:
ご指摘頂いた部分を修正しました。ご確認のほどお願いいたします。

https://github.com/ars-ishi/chat-space/pull/11#discussion_r219785414
